include/Beman/Optional26/detail/iterator.hpp

```diff
@@ -0,0 +1,147 @@
+// include/Beman/Optional26/detail/iterator.hpp                                            -*-C++-*-
+#ifndef BEMAN_OPTIONAL26_OPTIONAL_DETAIL_ITERATOR_HPP
+#define BEMAN_OPTIONAL26_OPTIONAL_DETAIL_ITERATOR_HPP
+
+#include <iterator>
+#include <type_traits>
+
+namespace beman::optional::detail {
+// Implementation inspired by libstdc++'s normal_iterator
+// (https://github.com/gcc-mirror/gcc/blob/master/libstdc%2B%2B-v3/include/bits/stl_iterator.h)
```

Comment:
Some of this code looks exactly like libstdc++'s class. libstdc++'s copyright and license should be reflected in this file.

Reply:
Hrm, and libstdc++'s license is not a Beman approved license.

Reply:
I would suggest using boost.iterator or stl_interfaces (this needs c++20 I believe. https://github.com/boostorg/stl_interfaces

Note that stl_interfaces is proposed for c++26 as well so this would be a good use case for it.

Reply:
@JeffGarland , pleaze check issue linked in the PR description

Reply:
Ok. I offered to help @tzlaine to important his iterator to Beman. I understood the wording may change.

Reply:
Well I would hope the wording would be the same, but don't stop!

Reply:
@camio , @JeffGarland , I made an experiment directly using `boost/stl_interfaces` as an external dep to `Beman.Optional26`. I was able to build and test with clang-18 (not sure yet why gcc13/14 won't work). 

Diff can be seen in https://github.com/neatudarius/Beman.Optional26/pull/2/.

Is this direction OK for everyody?

Reply:
Hrm, it'd be great to not pull in a huge Boost dependency. Is it feasible to put a copy of that code in the detail namespace instead?

Reply:
I copied 3 files from that Boost library in a6c6fa3bae9367733ebf354f4b7ae24c7da6965b

```shell
tree include/Beman/Optional26/detail 
include/Beman/Optional26/detail
├── iterator.hpp
└── stl_interfaces
    ├── config.hpp
    ├── fwd.hpp
    └── iterator_interface.hpp
 ```
 
 Next commit (f4eaa880ea6397907dc3636f215c69dc3b2c7189) renames the namespace / macros to be specific to `beman::optional::detail`.
 
 @camio , can you take a look?

Reply:
The better convention here is probably to put the stl_interfaces under external in the repository, but this is probably fine as well.  Ideally any macros / namespaces needed in stl interfaces would go back into stl_interfaces primary repo so that in the end a fetch content cmake would be able to grab the dependency instead of copying here.  I took a brief look at the iterator and it looked fine, but will try to dig deeper later.

So you're spearheading a number of technical things for Beman here -- I expect to spend some time on Thursday night discussing this project at the Beman night session.  Good stuff :)

Reply:
We discussed 2 approaches in this thread:

1. I've initially attached a commit from other branch were I had `external/stl_interfaces` - check 7457c76f9254b8ea494c3d6e790865c15f54b2d2. In this approach, I tried not change at all the Boost code (used `extern/stl_interfaces/include` as include directory in `src/Beman/Optional26/CMakeLists.txt`, and was able to directly include headers from `boost/stl_interfaces` in `iterator.hpp` in this repo).  

2. @camio  suggested that we better copy 3 headers from `boost/stl_interfaces` in `include/Beman/Optional26/detail` and then we rename whatever means boost names (namespaces, macros etc) to be sure that we don't have any interaction with other boost library versions installed on target system - basically, make `Beman.Optionak26` a standalone library, no production deps. When we'll have a `Beman.STLInterfaces` library available we can remove copied files.

> The better convention here is probably to put the stl_interfaces under external in the repository, but this is probably fine as well. Ideally any macros / namespaces needed in stl interfaces would go back into stl_interfaces primary repo so that in the end a fetch content cmake would be able to grab the dependency instead of copying here. 

@JeffGarland , I have a clarifying question here: are you aiming for option1 or option 2 or a mixt of them/new option 3? The `fetch content` suggestion seems to be `option 2` with no extra changes :D

Reply:
For now, I'm basically fine with either -- but I think our long term target should be fetch content with Beman stuff incorporated into the Boost source repo. Unless it is @tzlaine vision to port stl_interfaces to Beman specific repo -- but I don't see that as being needed. I very much prefer the external location (copied or fetched) since that makes it clear that this is work from elsewhere that is technically incidental to the implementation here. I certainly expect this won't be our last Boost dependency -- still many useful tools over there for library building.

Reply:
> So you're spearheading a number of technical things for Beman here -- I expect to spend some time on Thursday night discussing this project at the Beman night session. Good stuff :)

Do we have something on agenda? :D please share a link.

Reply:
https://wiki.edg.com/bin/view/Wg21stlouis2024/EveningSessions

Are you going to be there?

Reply:
ok. that's 3:30 AM Bucharest time :( I'll try...

Reply:
Closed this thread as now we have a copy of boost/stl_interfaces sources + permission to use them under "Apache-2.0 WITH LLVM-exception"  licence (https://github.com/beman-project/Optional26/pull/18#discussion_r1651196265).